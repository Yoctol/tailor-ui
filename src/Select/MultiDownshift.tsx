import Downshift, { DownshiftInterface } from 'downshift3';
import React, { Component, ReactNode } from 'react';

import { itemToString } from './utils';

type Option =
  | {
      label: string;
      value: string;
    }
  | string
  | number;

type MultiDownshiftProps = DownshiftInterface<Option> & {
  selectedItem?: Option[];
  initialSelectedItem?: Option[];
  onSelect?: (options: Option | Option[]) => void;
  onChange?: (options: Option | Option[]) => void;
  children: (options: any) => ReactNode;
};

interface MultiDownshiftState {
  selectedItems: Option[];
}

class MultiDownshift extends Component<
  MultiDownshiftProps,
  MultiDownshiftState
> {
  state: MultiDownshiftState = {
    selectedItems:
      this.props.initialSelectedItem || this.props.selectedItem || [],
  };

  get selectedItems() {
    const { selectedItem } = this.props;

    if (selectedItem) {
      return selectedItem;
    }

    return this.state.selectedItems;
  }

  stateReducer = (state: any, changes: any) => {
    switch (changes.type) {
      case Downshift.stateChangeTypes.blurButton:
        return {};
      case Downshift.stateChangeTypes.keyDownEnter:
      case Downshift.stateChangeTypes.mouseUp:
      case Downshift.stateChangeTypes.clickItem:
        return {
          ...changes,
          highlightedIndex: state.highlightedIndex,
          isOpen: true,
          inputValue: '',
        };
      default:
        return changes;
    }
  };

  handleChange = () => {
    const { onSelect, onChange } = this.props;
    const { selectedItems } = this.state;

    if (onSelect) {
      onSelect(selectedItems);
    }
    if (onChange) {
      onChange(selectedItems);
    }
  };

  handleSelection = (selectedItem: Option) => {
    const { onSelect, onChange } = this.props;

    if ((selectedItem as any).label === 'CREATE_OPTION') {
      if (onSelect) {
        onSelect(selectedItem);
      }
      if (onChange) {
        onChange(selectedItem);
      }

      return;
    }

    if (
      this.state.selectedItems
        .map(itemToString)
        .includes(itemToString(selectedItem))
    ) {
      this.removeItem(selectedItem);
    } else {
      this.addSelectedItem(selectedItem);
    }
  };

  removeItem = (item: Option) => {
    this.setState(
      {
        selectedItems: this.selectedItems.filter(
          (i) => itemToString(i) !== itemToString(item)
        ),
      },
      this.handleChange
    );
  };

  addSelectedItem(item: Option) {
    this.setState(
      {
        selectedItems: [...this.selectedItems, item],
      },
      this.handleChange
    );
  }

  getRemoveButtonProps = ({
    onClick,
    item,
    ...props
  }: {
    onClick?: (e: any) => void;
    item: Option;
  }) => {
    return {
      onClick: (e: any) => {
        // TODO: use something like downshift's composeEventHandlers utility instead
        if (onClick) {
          onClick(e);
        }
        e.stopPropagation();
        this.removeItem(item);
      },
      ...props,
    };
  };

  render() {
    const { children, ...props } = this.props;
    // TODO: compose together props (rather than overwriting them) like downshift does
    return (
      <Downshift
        {...props}
        stateReducer={this.stateReducer}
        onChange={this.handleSelection}
        selectedItem={null}
      >
        {(downshift) =>
          children({
            getRemoveButtonProps: this.getRemoveButtonProps,
            removeItem: this.removeItem,
            selectedItems: this.selectedItems,
            ...downshift,
          })
        }
      </Downshift>
    );
  }
}

export default MultiDownshift;
