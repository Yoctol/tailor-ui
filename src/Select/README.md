# Select Docs

```js
import { Select } from '@yoctol/ui';
```

### Try it with `knobs`:

<!-- STORY -->

## API

This component is extends from `react-select@2`, so please checkout the component's API at [the website](https://deploy-preview-2289--react-select.netlify.com/props).
