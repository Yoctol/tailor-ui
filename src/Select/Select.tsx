import BaseSelect, { components } from 'react-select';
import CreatableSelect from 'react-select/lib/Creatable';
import React, { FunctionComponent } from 'react';
import { MdSearch } from 'react-icons/md';

import styled from 'utils/styled-components';

import Box from '../Grid/Box';
import Flex from '../Grid/Flex';
import Icon from '../Icon';

const getStyledSelect = (creatable: boolean) => {
  const SelectComponent = creatable ? CreatableSelect : BaseSelect;

  return styled<any>(SelectComponent)`
    & .yoctol-select__control {
      min-width: 150px;
      border-color: ${p => p.theme.colors.gray300};
      background-color: ${p => p.theme.colors.light};
      box-shadow: none;

      &:hover {
        border-color: ${p => p.theme.colors.primary};
      }

      &.yoctol-select__control--is-focused {
        border-color: ${p => p.theme.colors.primaryDark};
      }

      &.yoctol-select__control-is-disabled {
        & .yoctol-select__indicators {
          &::before {
            opacity: 0.5;
          }
        }
      }
    }

    & .yoctol-select__menu {
      margin-top: 2px;

      .yoctol-select__menu-list {
        padding-top: 0;
        padding-bottom: 0;

        .yoctol-select__option {
          &:not(:last-child) {
            border-bottom: ${p => p.theme.borders.base}
              ${p => p.theme.colors.gray300};
          }
          background-color: ${p => p.theme.colors.light};
          color: ${p => p.theme.colors.gray700};

          &.yoctol-select__option--is-focused {
            background-color: ${p => p.theme.colors.primaryDark};
            color: ${p => p.theme.colors.light};
          }
        }
      }
    }
  `;
};

export interface ISelectProps {
  /**
   * delimiter string for multi-select value in text style mode
   */
  delimiter?: string;
  /**
   * Is the select value clearable
   */
  isClearable?: boolean;
  /**
   * Allow create new option
   */
  creatable?: boolean;
  /**
   * Disable the control
   */
  isDisabled?: boolean;
  /**
   * Allow user to select multiple options
   */
  isMulti?: boolean;
  /**
   * Allow the user to search for matching options
   */
  isSearchable?: boolean;
  /**
   * Choose tag or text style in multi-select (options: "tag", "text")
   */
  multiSelectMode?: string;
  /**
   * Specify the options the user can select from
   */
  options?: number[] | string[] | object[];
  /**
   * Change the text displayed when no option is selected
   */
  placeholder?: string;
  /**
   * One of options
   */
  value?: number | string | object | object[];
  /**
   * Subscribe to change events
   */
  onChange?: (option: number | string | object) => void;
}

const ValueContainer: FunctionComponent<any> = ({ children, ...props }) => (
  <components.ValueContainer {...props}>
    <Flex width="100%" alignItems="center">
      <Icon type={MdSearch} size="20" mr="2" />
      <Box flex="auto" position="relative">
        {children}
      </Box>
    </Flex>
  </components.ValueContainer>
);

const Select: FunctionComponent<ISelectProps> = props => {
  const { isMulti, delimiter, multiSelectMode, creatable = false } = props;

  const StyledSelect = getStyledSelect(creatable);

  const isSearchable = creatable || props.isSearchable;

  let selectProps: any = {
    classNamePrefix: 'yoctol-select',
    createOptionPosition: 'first',
    ...props,
    isSearchable,
  };

  if (isSearchable) {
    selectProps.components = { ValueContainer };
  }

  if (isMulti) {
    if (multiSelectMode === 'text') {
      const Option: FunctionComponent<any> = _props => (
        <components.Option {..._props} isFocused={_props.isSelected} />
      );

      const MultiValue = ({
        selectProps: {
          value: [{ value: firstValue }],
        },
        data: { value: dataValue, label: dataLabel },
      }: any) => `${firstValue !== dataValue ? delimiter : ''}${dataLabel}`;

      selectProps = {
        ...selectProps,
        components: {
          ...selectProps.components,
          Option,
          MultiValue,
        },
        closeMenuOnSelect: false,
        hideSelectedOptions: false,
      };
    }
  }

  return <StyledSelect {...selectProps} />;
};

Select.defaultProps = {
  delimiter: ',',
  isClearable: false,
  creatable: false,
  isDisabled: false,
  isMulti: false,
  isSearchable: false,
  multiSelectMode: 'tag',
  placeholder: '',
};

export default Select;
