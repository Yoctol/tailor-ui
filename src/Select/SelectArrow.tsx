import React, { FC } from 'react';
import { MdKeyboardArrowUp } from 'react-icons/md';
import { animated, config, to, useSpring } from '@react-spring/web';

import { Icon } from '../Icon';

const SelectArrow: FC<{ on: boolean }> = ({ on }) => {
  const { deg } = useSpring({
    deg: on ? 0 : 180,
    config: config.stiff,
  });

  return (
    <animated.i
      style={{
        transform: to(deg, (d) => `rotate(${d}deg)`),
        flex: 'none',
        pointerEvents: 'none',
        display: 'inline-flex',
      }}
    >
      <Icon type={MdKeyboardArrowUp} fill="gray400" />
    </animated.i>
  );
};

export default SelectArrow;
