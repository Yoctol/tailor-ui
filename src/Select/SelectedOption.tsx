import React, { FC } from 'react';
import { MdClose } from 'react-icons/md';

import { Icon } from '../Icon';

import { StyledSelectedOption } from './styles';
import { getDataTestId } from './utils';

export interface SelectedOption {
  index: number;
  size: 'sm' | 'md' | 'lg';
  removeButtonProps: Record<string, any>;
  'data-testid'?: string;
}

const SelectedOption: FC<SelectedOption> = ({
  children,
  index,
  removeButtonProps,
  ...props
}) => {
  return (
    <StyledSelectedOption
      {...props}
      {...getDataTestId(props, `selected-option-${index}`)}
    >
      {children}
      <Icon
        type={MdClose}
        size="16"
        cursor="pointer"
        ml="1"
        {...removeButtonProps}
        {...getDataTestId(props, `selected-option-${index}-clear-icon`)}
      />
    </StyledSelectedOption>
  );
};

export default SelectedOption;
