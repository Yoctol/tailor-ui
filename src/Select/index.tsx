import BaseSelect from 'react-select';
import React, { SFC } from 'react';

import styled from 'utils/styled-components';

const StyledSelect = styled<any, any>(BaseSelect)`
  & .yoctol-select__control {
    min-width: 150px;
    border-color: ${p => p.theme.colors.gray[8]};
    background-color: ${p => p.theme.colors.light};
    box-shadow: none;

    &:hover {
      border-color: ${p => p.theme.colors.primary};
    }

    &.yoctol-select__control--is-focused {
      border-color: ${p => p.theme.colors.primaryDark};
    }

    &.yoctol-select__control-is-disabled {
      & .yoctol-select__indicators {
        &::before {
          opacity: 0.5;
        }
      }
    }
  }

  & .yoctol-select__menu {
    margin-top: 2px;

    .yoctol-select__menu-list {
      padding-top: 0;
      padding-bottom: 0;

      .yoctol-select__option {
        &:not(:last-child) {
          border-bottom: ${p => p.theme.borders.base}
            ${p => p.theme.colors.gray[8]};
        }
        background-color: ${p => p.theme.colors.light};
        color: ${p => p.theme.colors.gray[2]};

        &.yoctol-select__option--is-focused {
          background-color: ${p => p.theme.colors.primaryDark};
          color: ${p => p.theme.colors.light};
        }
      }
    }
  }
`;

export interface SelectProps {
  /**
   * Is the select value clearable
   */
  isClearable?: boolean;
  /**
   * Disable the control
   */
  isDisabled?: boolean;
  /**
   * Allow the user to search for matching options
   */
  isSearchable?: boolean;
  /**
   * Specify the options the user can select from
   */
  options?: number[] | string[] | object[];
  /**
   * Change the text displayed when no option is selected
   */
  placeholder?: string;
  /**
   * One of options
   */
  value?: number | string | object;
  /**
   * Subscribe to change events
   */
  onChange?: (option: number | string | object) => void;
}

const Select: SFC<SelectProps> = props => (
  <StyledSelect classNamePrefix="yoctol-select" {...props} />
);

export default Select;
