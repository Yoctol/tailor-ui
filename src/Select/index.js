import Select from 'react-select';
import 'react-select/dist/react-select.css';

export default Select;
