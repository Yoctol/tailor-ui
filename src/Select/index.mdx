---
name: Select
menu: Data Entry
route: /data-entry/select
---

import { PropsTable, Playground } from 'docz';
import { Value } from 'react-powerplug';
import { range } from 'ramda';
import Select from './';

# Select

## Properties

<PropsTable of={Select} />

## Usage

```js
import { Select } from '@yoctol/ui';
```

### Basic

<Playground>
  <Value initial={{ label: 'Item 1', value: 1 }}>
    {({ value, set }) => (
      <Select
        value={value}
        onChange={set}
        options={[
          {
            label: 'Item 1',
            value: 1
          },
          {
            label: 'Item 2',
            value: 2
          },
          {
            label: 'Item 3',
            value: 3
          },
          {
            label: 'Item 4',
            value: 4
          },
          {
            label: 'Item 5',
            value: 5
          }
        ]}
      />
    )}
  </Value>
</Playground>

### isDisabled

<Playground>
  <Value initial={{ label: 'Item 1', value: 1 }}>
    {({ value, set }) => (
      <Select
        isDisabled
        value={value}
        onChange={set}
        options={[
          {
            label: 'Item 1',
            value: 1
          },
          {
            label: 'Item 2',
            value: 2
          },
          {
            label: 'Item 3',
            value: 3
          },
          {
            label: 'Item 4',
            value: 4
          },
          {
            label: 'Item 5',
            value: 5
          }
        ]}
      />
    )}
  </Value>
</Playground>

## With placeholder & isClearable

<Playground>
  <Value>
    {({ value, set }) => (
      <Select
        isClearable
        placeholder="どれ"
        value={value}
        onChange={set}
        options={[
          {
            label: 'Item 1',
            value: 1
          },
          {
            label: 'Item 2',
            value: 2
          },
          {
            label: 'Item 3',
            value: 3
          },
          {
            label: 'Item 4',
            value: 4
          },
          {
            label: 'Item 5',
            value: 5
          }
        ]}
      />
    )}
  </Value>
</Playground>
