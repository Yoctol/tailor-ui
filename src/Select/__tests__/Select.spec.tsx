import React, { useState } from 'react';
import userEvent from '@testing-library/user-event';

import { render, screen } from 'test/test-utils';

import { Select } from '../Select';

const DEFAULT_OPTIONS = [
  { label: 'Banana', value: 'banana' },
  { label: 'Orange', value: 'orange' },
  { label: 'Apple', value: 'apple' },
  { label: 'Mango', value: 'mango' },
];

const DEFAULT_SELECT_OPTION = { label: 'Banana', value: 'banana' };

const ControlSelect = (props: Record<string, any>) => {
  const [value, setValue] = useState(DEFAULT_SELECT_OPTION);

  return (
    <Select
      value={value}
      onChange={(newValue) => setValue(newValue)}
      options={DEFAULT_OPTIONS}
      {...props}
    />
  );
};

describe('Select', () => {
  it('should render correctly', () => {
    const { baseElement } = render(<Select options={DEFAULT_OPTIONS} />);

    expect(baseElement).toMatchSnapshot();
  });

  it('should not open menu when disabled', () => {
    render(<ControlSelect data-testid="select" disabled />);

    const select = screen.getByTestId('select');
    userEvent.click(select);

    expect(screen.queryByTestId('select-menu')).toBeNull();
  });

  it('should not open menu when loading', () => {
    render(<ControlSelect data-testid="select" loading />);

    const select = screen.getByTestId('select');
    userEvent.click(select);

    expect(screen.queryByTestId('select-menu')).toBeNull();
  });

  it.skip('should change value correctly when click option', async () => {
    render(<ControlSelect data-testid="select" />);

    const input = screen.getByTestId('select-input') as HTMLInputElement;
    expect(input.value).toBe('Banana');

    const select = screen.getByTestId('select');
    userEvent.click(select);

    const orange = screen.getByTestId('select-item-1');
    userEvent.click(orange);

    expect(input.value).toBe('Orange');
  });

  describe.skip('searchable', () => {
    it('should search correctly when change select input', async () => {
      render(<ControlSelect searchable data-testid="select" />);

      const input = screen.getByTestId('select-input') as HTMLInputElement;
      expect(input.value).toBe('Banana');

      const select = screen.getByTestId('select');
      userEvent.click(select);

      expect(input.value).toBe('');
      expect(input.placeholder).toBe('Banana');

      userEvent.type(input, 'Orange');

      const orange = screen.getByTestId('select-item-0');
      userEvent.click(orange);

      expect(screen.getByTestId('select-item-0')).toHaveTextContent('Orange');
      expect(screen.queryByTestId('select-item-1')).toBeNull();
      expect(screen.queryByText('Banana')).toBeNull();
    });
  });

  describe.skip('creatable', () => {
    it('should create successfully when pass creatable', async () => {
      const CreatableSelect = () => {
        const [loading, setLoading] = useState(false);
        const [value, setValue] = useState(DEFAULT_SELECT_OPTION);
        const [options, setOptions] = useState(DEFAULT_OPTIONS);

        return (
          <Select
            creatable
            data-testid="select"
            loading={loading}
            value={value}
            onChange={(newValue) => setValue(newValue)}
            options={options}
            isValidNewOption={(name) =>
              !options.map((option) => option.value).includes(name) &&
              name.trim() !== ''
            }
            onCreateOption={(name) => {
              const newOption = { label: name, value: name };
              setLoading(true);
              setTimeout(() => {
                setOptions([...options, newOption]);
                setValue(newOption);
                setLoading(false);
              }, 1000);
            }}
          />
        );
      };

      jest.useFakeTimers();

      render(<CreatableSelect />);

      const input = screen.getByTestId('select-input') as HTMLInputElement;
      expect(input.value).toBe('Banana');

      const select = screen.getByTestId('select');
      userEvent.click(select);

      userEvent.type(input, 'XXXXXX');

      const newOption = screen.getByText('Create new option: XXXXXX');
      expect(newOption).toBeInTheDocument();
      userEvent.click(newOption);

      expect(screen.getByTitle('loading')).toBeInTheDocument();

      jest.runOnlyPendingTimers();

      expect(screen.queryByTitle('loading')).toBeNull();
      expect(input.value).toBe('XXXXXX');

      jest.useRealTimers();
    });

    it('should not creatable when option is not valid', async () => {
      const CreatableSelect = () => {
        const [value, setValue] = useState(DEFAULT_SELECT_OPTION);

        return (
          <Select
            creatable
            data-testid="select"
            value={value}
            onChange={(newValue) => setValue(newValue)}
            options={DEFAULT_OPTIONS}
            isValidNewOption={(name) => name === 'NOT_VALID'}
          />
        );
      };

      render(<CreatableSelect />);

      const input = screen.getByTestId('select-input') as HTMLInputElement;
      expect(input.value).toBe('Banana');

      const select = screen.getByTestId('select');
      userEvent.click(select);

      userEvent.type(input, 'NOT_VALID');

      const newOption = screen.queryByText('Create new option: NOT_VALID');
      expect(newOption).toBeInTheDocument();
    });
  });

  describe('clearable', () => {
    it('should clear value when click clear icon', async () => {
      const ClearableSelect = () => {
        const [value, setValue] = useState(DEFAULT_SELECT_OPTION);

        return (
          <Select
            clearable
            data-testid="select"
            value={value}
            onChange={(newValue) => setValue(newValue)}
            options={DEFAULT_OPTIONS}
          />
        );
      };

      render(<ClearableSelect />);

      const input = screen.getByTestId('select-input') as HTMLInputElement;
      expect(input.value).toBe('Banana');

      const clearIcon = screen.getByTestId('select-clear-icon');
      userEvent.click(clearIcon);

      expect(input.value).toBe('');
    });
  });

  describe('multiple', () => {
    const MultipleSelect = () => {
      const [loading, setLoading] = useState(false);
      const [value, setValue] = useState([
        { label: 'Banana', value: 'banana' },
        { label: 'Orange', value: 'orange' },
      ]);
      const [options, setOptions] = useState([
        { label: 'Banana', value: 'banana' },
        { label: 'Orange', value: 'orange' },
        { label: 'Apple', value: 'apple' },
        { label: 'Mango', value: 'mango' },
      ]);

      return (
        <Select
          creatable
          data-testid="select"
          loading={loading}
          width="360px"
          multiple
          value={value}
          onChange={(newValue) => setValue(newValue)}
          options={options}
          isValidNewOption={(name) =>
            !options.map((option) => option.label).includes(name) &&
            name.trim() !== ''
          }
          onCreateOption={(name) => {
            const newOption = { label: name, value: name };
            setLoading(true);
            setTimeout(() => {
              setOptions([...options, newOption]);
              setValue((prevValue) => [...prevValue, newOption]);
              setLoading(false);
            }, 1000);
          }}
        />
      );
    };

    it('should display multi option', () => {
      const { baseElement } = render(<MultipleSelect />);

      expect(baseElement).toMatchSnapshot();

      expect(screen.getByText('Banana')).toBeInTheDocument();
      expect(screen.getByText('Orange')).toBeInTheDocument();
    });

    it.skip('should add selected option when click other option', async () => {
      render(<MultipleSelect />);

      const select = screen.getByTestId('select');
      userEvent.click(select);

      userEvent.click(screen.getByText('Apple'));

      expect(screen.getByTestId('select-menu')).toBeInTheDocument();

      userEvent.click(screen.getByText('Mango'));

      userEvent.click(select);

      expect(screen.queryByTestId('select-menu')).toBeNull();
      expect(screen.getByText('Apple')).toBeInTheDocument();
      expect(screen.getByText('Mango')).toBeInTheDocument();
    });

    it('should remove selected option when click clear ', async () => {
      render(<MultipleSelect />);

      userEvent.click(
        screen.getByTestId('select-selected-option-1-clear-icon')
      );
      userEvent.click(
        screen.getByTestId('select-selected-option-0-clear-icon')
      );

      expect(screen.queryByText('Banana')).toBeNull();
      expect(screen.queryByText('Orange')).toBeNull();
    });

    it.skip('should remove last selected option when press backspace ', async () => {
      render(<MultipleSelect />);

      const select = screen.getByTestId('select');
      userEvent.click(select);

      const input = screen.getByTestId('select-input') as HTMLInputElement;
      userEvent.type(input, '{backspace}');
      userEvent.click(select);

      expect(screen.queryByTestId('select-menu')).toBeNull();
      expect(screen.getByText('Banana')).toBeInTheDocument();
      expect(screen.queryByText('Orange')).toBeNull();
    });
  });
});
