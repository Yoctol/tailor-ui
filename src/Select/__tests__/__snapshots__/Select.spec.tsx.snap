// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select multiple should display multi option 1`] = `
<body>
  <div
    portal-container=""
  >
    .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  top: 72px;
  right: 10px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
      style="position: fixed; top: 0px; left: 0px; z-index: 15;"
    >
      <div
        class="c0"
      />
    </div>
    .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  right: 10px;
  bottom: 10px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
      style="position: fixed; top: 0px; left: 0px; z-index: 30;"
    >
      <div
        class="c0"
      />
    </div>
  </div>
  .c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
}

.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  line-height: 1;
  cursor: pointer;
  margin-left: 0.25rem;
}

.c4 svg {
  fill: #67758d;
  vertical-align: middle;
  pointer-events: none;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c4.c4.c4 svg {
  width: 16px;
  height: 16px;
}

.c6 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  line-height: 1;
  cursor: default;
}

.c6 svg {
  fill: #67758d;
  vertical-align: middle;
  pointer-events: none;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c6.c6.c6 svg {
  fill: #97a4ba;
  width: 24px;
  height: 24px;
}

.c5 {
  visibility: hidden;
  position: absolute;
  pointer-events: none;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  cursor: pointer;
}

.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: 100%;
  margin-right: 8px;
  padding: 0 8px;
  border: 1px solid;
  border-radius: 0.25rem;
  border-color: #3990cd;
  background-color: #fbfbfd;
  color: #3990cd;
  font-size: 0.875rem;
  min-height: calc(2.25rem - 10px);
  margin-top: 2px;
  margin-bottom: 2px;
}

.c3 > span {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c3 svg {
  fill: #a2d3ff;
}

.c3:hover svg {
  fill: #3990cd;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  max-width: 100%;
  border: 1px solid;
  border-radius: 0.25rem;
  border-color: #97a4ba;
  outline: none;
  background-color: #ffffff;
  color: #283040;
  line-height: 1.5;
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  min-height: 2.25rem;
  padding: 2px 0.75rem;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c1:hover {
  border-color: #328ef5;
}

.c1 input {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  width: 100%;
  overflow: hidden;
  border: none;
  outline: none;
  background-color: transparent;
  text-overflow: ellipsis;
  pointer-events: none;
}

.c1 input::-webkit-input-placeholder {
  color: #97a4ba;
}

.c1 input::-moz-placeholder {
  color: #97a4ba;
}

.c1 input:-ms-input-placeholder {
  color: #97a4ba;
}

.c1 input::placeholder {
  color: #97a4ba;
}

.c1 input {
  font-size: 1rem;
}

<div>
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="downshift-4-label"
      class="c0"
      role="combobox"
      style="width: 360px;"
    >
      <div
        class="c1"
        data-testid="select"
        id="downshift-4"
      >
        <div
          class="c2"
          overflow="hidden"
        >
          <div
            class="c3"
            data-testid="select-selected-option-0"
          >
            Banana
            <i
              class="c4"
              cursor="pointer"
              data-testid="select-selected-option-0-clear-icon"
              size="16"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </i>
          </div>
          <div
            class="c3"
            data-testid="select-selected-option-1"
          >
            Orange
            <i
              class="c4"
              cursor="pointer"
              data-testid="select-selected-option-1-clear-icon"
              size="16"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </i>
          </div>
          <input
            aria-autocomplete="list"
            aria-labelledby="downshift-4-label"
            autocomplete="off"
            data-testid="select-input"
            id="downshift-4-input"
            placeholder=""
            style="width: 18px; box-sizing: content-box;"
            value=""
          />
          <span
            class="c5"
            style="font-size: 16px;"
          />
        </div>
        <i
          style="transform: rotate(180deg); flex: 0 0 auto; pointer-events: none; display: inline-flex;"
        >
          <i
            class="c6"
            cursor="default"
            fill="gray400"
            size="24"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z"
              />
            </svg>
          </i>
        </i>
      </div>
    </div>
  </div>
</body>
`;

exports[`Select should render correctly 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
}

.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  line-height: 1;
  cursor: default;
}

.c4 svg {
  fill: #67758d;
  vertical-align: middle;
  pointer-events: none;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c4.c4.c4 svg {
  fill: #97a4ba;
  width: 24px;
  height: 24px;
}

.c3 {
  visibility: hidden;
  position: absolute;
  pointer-events: none;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  right: 10px;
  bottom: 10px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  top: 72px;
  right: 10px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  cursor: pointer;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  max-width: 100%;
  border: 1px solid;
  border-radius: 0.25rem;
  border-color: #97a4ba;
  outline: none;
  background-color: #ffffff;
  color: #283040;
  line-height: 1.5;
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  min-height: 2.25rem;
  padding: 4px 0.75rem;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c1:hover {
  border-color: #328ef5;
}

.c1 input {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  width: 100%;
  overflow: hidden;
  border: none;
  outline: none;
  background-color: transparent;
  text-overflow: ellipsis;
  pointer-events: none;
}

.c1 input::-webkit-input-placeholder {
  color: #97a4ba;
}

.c1 input::-moz-placeholder {
  color: #97a4ba;
}

.c1 input:-ms-input-placeholder {
  color: #97a4ba;
}

.c1 input::placeholder {
  color: #97a4ba;
}

.c1 input {
  font-size: 1rem;
}

<body>
  <div>
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="downshift-0-label"
      class="c0"
      role="combobox"
      style="width: 240px;"
    >
      <div
        class="c1"
        id="downshift-0"
      >
        <div
          class="c2"
          overflow="hidden"
        >
          <input
            aria-autocomplete="list"
            aria-labelledby="downshift-0-label"
            autocomplete="off"
            id="downshift-0-input"
            placeholder=""
            style="width: 18px; box-sizing: content-box;"
            value=""
          />
          <span
            class="c3"
            style="font-size: 16px;"
          />
        </div>
        <i
          style="transform: rotate(180deg); flex: 0 0 auto; pointer-events: none; display: inline-flex;"
        >
          <i
            class="c4"
            cursor="default"
            fill="gray400"
            size="24"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z"
              />
            </svg>
          </i>
        </i>
      </div>
    </div>
  </div>
  <div
    portal-container=""
  >
    <div
      style="position: fixed; top: 0px; left: 0px; z-index: 15;"
    >
      <div
        class="c5"
      />
    </div>
    <div
      style="position: fixed; top: 0px; left: 0px; z-index: 30;"
    >
      <div
        class="c6"
      />
    </div>
  </div>
</body>
`;
